src/dlfcn.c

```diff
@@ -467,7 +472,170 @@ char *dlerror( void )
+        }
+   }
+   return NULL;
+}
+
+static BOOL is_valid_address( void *addr )
+{
+    if( addr == NULL )
+        return FALSE;
+
+    /* check valid pointer */
+    MEMORY_BASIC_INFORMATION info;
+    SIZE_T result = VirtualQuery( addr, &info, sizeof( info ) );
+    if( result == 0 || info.AllocationBase == NULL || info.AllocationProtect == 0 || info.AllocationProtect == PAGE_NOACCESS )
+        return FALSE;
+    return TRUE;
+}
+
+/*
+ * Return state if address points to an import thunk
+ *
+ * An import thunk is setup with a 'jmp' instruction followed by an
+ * absolute address (32bit) or relative offset (64bit) pointing into
+ * the import address table (iat), which is partially maintained by
+ * the runtime linker.
+ */
+static BOOL is_import_thunk( void *addr )
+{
+    return *(short *)addr == 0x25ff ? TRUE : FALSE;
+}
+
+/*
+ * Return adress from the import address table (iat), if
+ * the original address points to a thunk table entry.
+ */
+static void *get_address_from_import_address_table( void *iat, DWORD iat_size, void *addr )
+{
+    BYTE *thkp = (BYTE*)addr;
+    /* get offset from thunk table (after instruction 0xff 0x25)
+     *   4018c8 <_VirtualQuery>: ff 25 4a 8a 00 00
+     */
+    ULONG offset = *(ULONG*)(thkp + 2);
+#ifdef _WIN64
+    /* On 64 bit the offset is relative
+     *   4018c8:	ff 25 4a 8a 00 00    	jmpq   *0x8a4a(%rip)        # 40a318 <__imp_VirtualQuery>
+     */
+    void *ptr = (void *)(thkp + 6 + offset);
+#else
+    /* On 32 bit the offset is absolute
+     *    4019b4:	ff 25 90 71 40 00    	jmp    *0x40719
+     */
+    void *ptr = (void *)offset;
+#endif
+
+    if( !is_valid_address( ptr ) ||  ptr < iat || ptr > iat + iat_size )
+        return NULL;
+
+    return *(void **)ptr;
+}
+
+/* holds module filename */
+static char module_filename[2*MAX_PATH];
+
+static BOOL fill_module_info( void *addr, Dl_info *info )
+{
+    HMODULE hModule;
+    DWORD dwSize;
+
+    info->dli_fname = NULL;
+    if( !GetModuleHandleExA( GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS | GET_MODULE_HANDLE_EX_FLAG_UNCHANGED_REFCOUNT, addr, &hModule ) || !hModule )
+    {
+        info->dli_fbase = NULL;
+        return FALSE;
+    }
+
+    info->dli_fbase = (void *) hModule;
+
+    dwSize = GetModuleFileNameA( hModule, module_filename, sizeof( module_filename ) );
+
+    if( dwSize == 0 || dwSize == sizeof( module_filename ) )
+        return FALSE;
+
+    info->dli_fname = module_filename;
+    return TRUE;
+}
+
+DLFCN_EXPORT
+int dladdr( void *addr, Dl_info *info )
+{
+    void *realAddr;
+    HMODULE hModule;
+    IMAGE_IMPORT_DESCRIPTOR *iat;
+    DWORD dwSize = 0;
+
+    if( !addr || !info )
+        return 0;
+
+    hModule = GetModuleHandleA( NULL );
+    if( !hModule )
+        return 0;
+
+    if( !is_valid_address( addr ) )
+        return 0;
+
+    realAddr = addr;
+    info->dli_sname = NULL;
+
+    iat = get_import_address_table( hModule, &dwSize );
+
+    if( is_import_thunk( addr ) ) {

```

Comment:
Maybe inline it? It is not used on any other place.

Reply:
I think this should remain a separate function, otherwise an additional comment would be required, what is this good for and it can be more easily expanded later if needed.